zeppelin and vyper with IMap

denisadiaconescu opened the pull request:
These changes were tested with the K and EVM commits specified in the .k.rev and .kevm.rev files.

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -71,56 +74,66 @@ requires:
     andBool VALUE >Int BAL_FROM
+ensures:
+    ensures M1 ==IMap M2 
+        except (SetItem(#hashedLocation({COMPILER}, {_BALANCES}, CALLER_ID)) .Set)
```
why do you need this `except` here? the original spec says that the storage does NOT update at all.

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -131,60 +144,122 @@ requires:
+ensures:
+    ensures select(M2, #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)) ==Int BAL_FROM -Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO +Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) ==Int ALLOW -Int VALUE
```
how could this possible where we have the hash collision?

daejunpark replied:
oh, BAL_FROM is equal to ALLOW

daejunpark replied:
i think it is better to explicitly say that BAL_FROM ==Int ALLOW in the ensures, to highlight it.

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -131,60 +144,122 @@ requires:
+ensures:
+    ensures select(M2, #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)) ==Int BAL_FROM -Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO -Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) ==Int ALLOW -Int VALUE
```
same here, BAL_TO ==Int ALLOW in ensures.

daejunpark replied:
not much sense to name it as success case.  you need to put a nice comment what this spec is trying to say.  without careful look, no one will even recognize BAL_TO -Int VALUE.

daejunpark replied:
not only for this one, but for all "success-1-*" cases.

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -131,60 +144,122 @@ requires:
+ensures:
+    ensures select(M2, #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) ==Int ALLOW -Int VALUE
+    andBool M1 ==IMap M2 except
+      (SetItem(#hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) .Set)
```
no hash collision here? why?

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -131,60 +144,122 @@ requires:
++requires:
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)) ==Int BAL_FROM
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO
+    andBool select(M1, #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID)) ==Int ALLOW
```
nitpick: follow the vertical alignment.  see the original spec for the storage.

denisadiaconescu replied:
I fixed this.

daejunpark replied:
no you didn't.  what did you fix?

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -131,60 +144,122 @@ requires:
+    andBool BAL_TO +Int VALUE <Int (2 ^Int 256)
+    andBool VALUE <=Int ALLOW
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) ==Int #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) ==Int #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)
```
how could the allowance location be the same with two different locations?

daejunpark replied:
i cannot understand these two cases: failure-1-[ab].
you are better to explain this nontrivial case split.

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -131,60 +144,122 @@ requires:
+ensures:
+    ensures M2 ==IMap M1 except 
+        (SetItem(#hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)) 
+         SetItem(#hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) .Set)
```
no hash collision stuff here?

daejunpark commented on erc20/zeppelin/zeppelin-erc20-spec.ini:
```diff
@@ -145,54 +159,112 @@ requires:
+[transferFrom-success-1-d]
++requires:
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) ==Int #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) ==Int #hashedLocation({COMPILER}, {_BALANCES}, TO_ID) 
```
i don't see how this is possible. we have a lemma saying that two hash2 terms are different if the arguments are different, and here we know that FROM_ID =/= TO_ID.

daejunpark commented on resources/lemmas.md:
```diff
@@ -173,6 +173,9 @@ It reduces the reasoning efforts of the underlying theorem prover, factoring out
     rule byteStack2IntList ( WS , 1 ) => #asWord ( WS [ 0 .. 32 ] ) .IntList
+
+    rule hash2(K1,V1) =/=Int hash2(K2,V2) => K1 =/=Int K2 orBool V1 =/=Int V2
+    rule hash2(K1,V1) ==Int hash2(K2,V2) => K1 ==Int K2 andBool V1 ==Int V2
```
this lemma does NOT make sense if you want to explicitly reason about the hash collision.

you have to be consistent, choosing either
- explicit reasoning about the hash collision without this lemma
- assuming that no hash collision with or without this lemma

daejunpark commented:
@denisadiaconescu overall looks ok, but i'm not happy with the hash collision reasoning.  i'm ok with explicit hash collision reasoning, but you should be consistent.  both explicit reasoning and having the hash2 lemma does NOT make sense.  see the detailed comments for each lines, and please address them.

denisadiaconescu commented:
Jenkins: test this please

denisadiaconescu commented:
@daejunpark Thanks for the review. I had the impression that I deleted the lemma for `hash2(K1,V1) ==Int hash2(K2,V2)`, sorry for this. Now I deleted the lemma, but I also dropped the partial hash collision analysis. 

The only major change now is in `transferFrom-success-1` (for both `vyper` and `zeppelin`) in which we need to add two extra conditions (that some hash values are different) in order to prove that case.

denisadiaconescu commented:
@daejunpark Thanks for the review. Sorry about that lemma, I had the impression that I deleted it and I was not trying to use it. Now I deleted the lemma, but also dropped the (partial) hash collision analysis. The main changes now are in `transferFrom-success-1` (for both `vyper` and `zeppelin`) in which we added two extra conditions (saying explicitly that some hash values are not equal) in order to prove that case. 

I cherry-picked the commit you pointed out, but Jenkins still fails with the error `No rule to make target 'jenkins'`.

daejunpark commented on Makefile:
```diff
@@ -97,8 +97,8 @@ zeppelin_erc20_files:=totalSupply-spec.k \
-             transferFrom-failure-2-spec.k
-
+             transferFrom-failure-2-spec.k 
+            
```
we shouldn't have this non-necessary white-space changes. please fix it.

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -71,56 +79,71 @@ requires:
+    M1
 +requires:
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, CALLER_ID)) ==Int BAL_FROM
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO
```
by alignment, i mean, align `==Int`, by adding some white spaces between `TO_ID))` and `==Int`.

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -71,56 +79,71 @@ requires:
+    andBool BAL_TO +Int VALUE <Int (2 ^Int 256)  
+ensures:
+    ensures select(M2, #hashedLocation({COMPILER}, {_BALANCES}, CALLER_ID)) ==Int BAL_FROM -Int VALUE
+    andBool select(M2, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO +Int VALUE
```
alignment here as well

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -71,56 +79,71 @@ requires:
+    M1
 +requires:
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, CALLER_ID)) ==Int BAL_FROM
+    andBool select(M1, #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)) ==Int BAL_TO
```
align

daejunpark commented on erc20/vyper/vyper-erc20-spec.ini:
```diff
@@ -131,60 +154,85 @@ requires:
     andBool BAL_TO +Int VALUE <Int (2 ^Int 256)
     andBool VALUE <=Int ALLOW
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) =/=Int #hashedLocation({COMPILER}, {_BALANCES}, FROM_ID)
+    andBool #hashedLocation({COMPILER}, {_ALLOWANCES}, FROM_ID CALLER_ID) =/=Int #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)
```
this doesn't look good.  you have to add this kind of requires for all specs, without using the lemma at all,  or no this kind of requires with using the lemma only.

denisadiaconescu commented:
Created new PR to `master` #99
